std-bits/table/src/main/java/org/enso/table/data/index/MultiValueIndex.java

```diff
@@ -127,22 +128,40 @@ public Table makeTable(Aggregator[] columns) {
         merged);
   }
 
-  public Table makeCrossTabTable(
+  public static Table makeCrossTabTable(
       Column[] groupingColumns,
       Column nameColumn,
       Aggregator[] aggregates,
       String[] aggregateNames) {
     Context context = Context.getCurrent();
     NameDeduplicator outputTableNameDeduplicator = new NameDeduplicator();
 
-    final int size = locs.size();
-
-    var nameIndex =
+    // Create combined index
+    Column combinedColumns[] =
+        Stream.concat(Arrays.stream(groupingColumns), Arrays.stream(new Column[] {nameColumn}))
+            .toArray(Column[]::new);
+    var combinedIndex =
         MultiValueIndex.makeUnorderedIndex(
-            new Column[] {nameColumn},
-            nameColumn.getSize(),
-            TextFoldingStrategy.unicodeNormalizedFold);
-    final int columnCount = groupingColumns.length + nameIndex.locs.size() * aggregates.length;
+            combinedColumns, nameColumn.getSize(), TextFoldingStrategy.unicodeNormalizedFold);
+
+    // Generate lists of combined keys and subkeys
+    List<UnorderedMultiValueKey> combinedKeys = new ArrayList<>(combinedIndex.keys());
+    List<UnorderedMultiValueKey> groupingSubkeys = new ArrayList<>(combinedKeys.size());
+    List<UnorderedMultiValueKey> nameSubKeys = new ArrayList<>(combinedKeys.size());
+    int[] groupingColumnIndices = IntStream.range(0, groupingColumns.length).toArray();
+    int[] nameColumnIndices = new int[] {groupingColumns.length};
+    for (var key : combinedKeys) {
+      groupingSubkeys.add(key.subKey(groupingColumnIndices));
+      nameSubKeys.add(key.subKey(nameColumnIndices));
+    }
+
+    // Fill numberings
+    ObjectNumberer<UnorderedMultiValueKey> nameNumberer = new ObjectNumberer<>(nameSubKeys);
+    ObjectNumberer<UnorderedMultiValueKey> groupingNumberer = new ObjectNumberer<>(groupingSubkeys);
```

Comment:
It feels like we could avoid concretely creating `groupingSubKeys` and `nameSubKeys`.

If the `ObjectNumberer` took a `Stream<>` in we could just map these through here.

Reply:
We later need to iterate over the combined, grouping, and name keys in parallel.